packages/babel-preset-default/README.md

```diff
@@ -2,6 +2,8 @@
 
 Default [Babel](https://babeljs.io/) preset for WordPress development.
 
+The preset includes configuration which enable language features and syntax extensions targeted for support by WordPress. This includes [ECMAScript proposals](https://github.com/tc39/proposals) which have reached [Stage 4 ("Finished")](https://tc39.es/process-document/), as well as the [JSX syntax extension](https://reactjs.org/docs/introducing-jsx.html). For more information, refer to the [JavaScript Coding Guidelines](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#javascript).
```

Comment:
We can also reference this section in Babel docs since we are using this flag to make sure stage 4 proposals are enabled before they become an official standard: https://babeljs.io/docs/en/babel-preset-env#shippedproposals

Reply:
I came to mention the `shippedProposals` bit: it's a bit confusing to me that we say here that only stage 4 proposals are allowed and then we enable lower stages through this flag. Docs are great, but if the compiler passes I'd assume my code is correct. A point in case: as of today, if I use numeric separators, how would I know it's not recommended (because it's stage 3 yet)?

I guess my point is that, if we want to enforce to use only stage 4 proposals, we should disable the `shippedProposals` for coherence. Otherwise, we'd also need to list them in the docs.

Reply:
Hm, I guess it's unexpected to me. I was under the impression we were targeting Stage 4, not "Shipped". In other words: We wouldn't be enabling any non-Stage 4 features. For me, I'd prefer that we don't. If it's unreasonably difficult to configure this, I'd be open to revising the commitment.

Just my opinion, of course.

Reply:
This flag covers **only** stage 4 features that aren’t part of an official document that is published each year. So it speeds up the process a bit knowing that those features were accepted and it’s certain that they will be included in the language specs soon.

Reply:
Documentation explains it this way:


> Note that this _does not_ enable the same transformations as `@babel/preset-stage-3`, since proposals can continue to change before landing in browsers.

Reply:
I think I must have misread the text, since my initial reading was that it included `stage-3` proposals and additional changes to said proposals that haven't yet reached the browser.

On a re-read (and your comment https://github.com/WordPress/gutenberg/pull/22030#discussion_r419504846), my new interpretation is that it's saying merely that it is including proposals which aren't yet in the _published_ standard, but which _have_ reached Stage 4. I think this is compatible with the text as originally proposed. Am I correct in that your original suggestion was to clarify how this works with Babel?

Reply:
Dug a bit into this and prepared a PR to revert the `shippedProposals` flag at https://github.com/WordPress/gutenberg/pull/22083

I also thought the idea was to only use stage-4 features (that are enabled by our babel preset) and not stage-3. I think it's clearer that we stick to stage-4, given how arbitrary the `shippedProposals` may be at any point in time (adding more confusion + we'll have to potentially revert some code changes if a stage-3 doesn't make it to stage-4). Having said that, this is a strong opinion loosely held: the important bit to me is that our docs & parser are in sync.

Reply:
Here is the list of features that get included:
https://github.com/babel/babel/blob/master/packages/babel-preset-env/data/shipped-proposals.js

Reply:
> Here is the list of features that get included:
> https://github.com/babel/babel/blob/master/packages/babel-preset-env/data/shipped-proposals.js

Hm, so it definitely does include non-Stage-4 proposals 🤔 At least the numeric separators mentioned previously, which is currently Stage 3:

https://github.com/tc39/proposal-numeric-separator

Reply:
It’s probably an exception, the up to date list with stage 4 is here:

https://github.com/tc39/proposals/blob/master/finished-proposals.md

Reply:
It's a bit convoluted to trace which things were included in `shippedProposals` at what stage. I've done a wee bit by looking at the historic babel [docs](https://babeljs.io/versions) & [releases](https://github.com/babel/babel/releases), as well as trying to match those with TC39 evolution.

It looks like `shippedProposals` was added in the v7 line (I hadn't found any mention before), so I was able to compile the whole list of things that were added. This is what I've found:

- `flatMap` entered [stage-4](https://github.com/tc39/proposal-flatMap/commit/d10a30b469f3d07423a7921b3f23ac42cdaa343d) in January 2019. I could trace babel support via `shippedProposals` to August 2018 ([release](https://babeljs.io/blog/2018/08/27/7.0.0#syntax), [docs](https://babeljs.io/docs/en/7.0.0/babel-preset-env#shippedproposals)) the earliest.

- `globalThis` entered [stage-4](https://github.com/tc39/proposal-global/commit/282d472a17ce404c4ef10982261e76f7341bddde) on October 2019. It was included as part of `shippedProposals` in [babel 7.4](https://babeljs.io/docs/en/7.4.0/babel-preset-env#shippedproposals) released in March 2019.

- `string.match-all` entered [stage-4](https://github.com/tc39/proposal-string-matchall/issues/21#issuecomment-476786525) in March 2019, and it was included as part of `shippedProposals` in [babel 7.4](https://babeljs.io/docs/en/7.4.0/babel-preset-env#shippedproposals) released in the same month.

- `numericSeparator` is still in [stage-3](https://github.com/tc39/proposal-numeric-separator), while it was included as part of `shippedProposals` in [babel 7.9](https://babeljs.io/docs/en/babel-preset-env#shippedproposals) released in March 2020.

It looks like `shippedProposals` is actually used to ship stage-3 proposals that, somehow, people _know/expect_ to become stage-4 in the next ~6 months.

Personally, I find that a bit confusing to explain and understand and would prefer only use [stage-4 features](https://github.com/WordPress/gutenberg/pull/22083). I only found [one case](https://github.com/WordPress/gutenberg/blob/36f420278e467ae7801bb9b691d321a274f14b42/bin/api-docs/update-api-docs.js#L171) in which we actually use the syntax enabled by `shippedProposals` (machAll), and it's a supporting node script, not even part of the Gutenberg plugin. All the dancing I had to do to understand speaks about the complexity this adds for the little benefit.

Although I have a preference, I don't feel super-strongly about it. If we decide to keep things as they are, we need to update docs accordingly. Perhaps the way forward here is to add a note about `shippedProposals` and a link to the babel release we use?

Reply:
I seem to recall that that browsers will often include new features before they've been deemed "finished", often for the purpose of getting real-world feedback. If this is what `shippedProposals` is aligning to, it's a totally different thing independent from the stage process. For example, in Chrome, I can use the Stage 3 numeric separator feature.

For me, I think it's a choice between:

- We continue to use `shippedProposals`, and change the support commitment to "features which have shipped in browsers".
   - Notably, no mention of stages, as they're totally independent, though often there's a correlation since browsers will often start shipping features only late in the proposal process.
- We stop using `shippedProposals` (#22083) and keep the support commitment to "Stage 4".

Reply:
I'd think that's what [stage 3](https://tc39.es/process-document/) is about: getting feedback from implementors.

Reply:
> I'd think that's what [stage 3](https://tc39.es/process-document/) is about: getting feedback from implementors.

Yep, it would appear so.

My current leaning: We're not a browser vendor, Stage 3 is still subject to changes, and our commitment to Stage 4 is in the name of stability. I don't think we should want `shippedProposals`, unless it's otherwise very difficult to configure to what we _do_ want (Stage 4). From #22083, it doesn't appear to be that big an issue.

It would be a breaking change for `@wordpress/babel-preset-default`, and might require some additional communication for "upstream" packages (`@wordpress/scripts`, etc), in case someone happened to be using a feature.

I'd be interested if @gziolo has thoughts, but I'm inclined to proceed with something like #22083.

Reply:
I was also digging into documents to figure out how the process of promotion from stage 3 to stage 4 looks like. I found out that 2 browser vendors need to implement the feature so it could advance to the final level. It's not clear from Babel documentation whether `shippedProposals` are guaranteed to become stage 4 soon.

@hzoo, would you mind helping us to solve this mystery? The whole discussion boils down to the question if the `shippedProposals` flag enables features that are guaranteed to be included in JavaScript specification :) For the WordPress project, we want to make sure that we never enable language features that might not get into the final spec.

Reply:
Dang, good to know the staging process and `shippedProposals` is still confusing, so we could be doing a better job of explaining everything.. 

>  The whole discussion boils down to the question if the shippedProposals flag enables features that are guaranteed to be included in JavaScript specification
quick point: If guaranteed means 100% then no, it's more like 90%. It means browsers like Chrome have implemented it and it's "possible" there will be tweaks but in most cases unlikely.

https://tc39.es/process-document/. shippedProposals is sorta like a compromise where it's like "3.5" - it's basically done and browsers want people to try it out but it's always possible to remove it or change it, especially if it's under a flag in the browser still. The difference between 3 and 4 is just that the feature is basically done but it depends on "real world" usage (although I'd argue that's what Babel is for) but you could talk about the difference between the API/syntax usage via Babel or the performance in native browsers, etc. Stage 4 is the guaranteed time if you want to be absolutely sure. In the end, a judgement call, just giving the option for what a lot of people are ok with. As an example, I Decorators were at Stage 3 and went down a stage, and then changed, class properties moved down and were split into different proposals, etc.

Reply:
Thank you for clarification. It looks like @nosolosw got it correct from start. I got confused by the following note in the Babel preset docs:
> Note that this does not enable the same transformations as `@babel/preset-stage-3`, since proposals can continue to change before landing in browsers.

It means we should disable `shippedProposals` flag as already proposed in https://github.com/WordPress/gutenberg/pull/22083.